File: trove/backend/internal/service/travel.go
package service

import (
	"context"
	"fmt"

	"github.com/google/uuid"
	"github.com/saifkazi/trove/backend/internal/dao"
	"github.com/saifkazi/trove/backend/internal/model"
)

type TravelService struct {
	store *dao.TravelStore
}

func (s *TravelService) ListLocations(ctx context.Context, search string) ([]model.TravelLocation, error) {
	return s.store.List(ctx, search)
}

func (s *TravelService) GetLocation(ctx context.Context, id string) (*model.TravelLocation, error) {
	uid, err := uuid.Parse(id)
	if err != nil {
		return nil, fmt.Errorf("invalid id: %w", err)
	}
	return s.store.Get(ctx, uid)
}

func (s *TravelService) CreateLocation(ctx context.Context, p dao.CreateLocationParams) (*model.TravelLocation, error) {
	return s.store.Create(ctx, p)
}

func (s *TravelService) UpdateLocation(ctx context.Context, p dao.UpdateLocationParams) (*model.TravelLocation, error) {
	return s.store.Update(ctx, p)
}

func (s *TravelService) DeleteLocation(ctx context.Context, id string) error {
	uid, err := uuid.Parse(id)
	if err != nil {
		return fmt.Errorf("invalid id: %w", err)
	}
	return s.store.Delete(ctx, uid)
}

func (s *TravelService) CreateSpot(ctx context.Context, p dao.CreateSpotParams) (*model.TouristSpot, error) {
	return s.store.CreateSpot(ctx, p)
}

func (s *TravelService) UpdateSpot(ctx context.Context, p dao.UpdateSpotParams) (*model.TouristSpot, error) {
	return s.store.UpdateSpot(ctx, p)
}

func (s *TravelService) DeleteSpot(ctx context.Context, id string) error {
	uid, err := uuid.Parse(id)
	if err != nil {
		return fmt.Errorf("invalid id: %w", err)
	}
	return s.store.DeleteSpot(ctx, uid)
}

File: trove/ui/src/services/travel.ts
import type { TravelLocation, TravelLocationFormData, TouristSpot } from '@/types/api'
import { rpc } from './client'

const SVC = 'trove.v1.travel.TravelService'

export const travelApi = {
  list: (params: { search?: string } = {}): Promise<TravelLocation[]> => {
    const body: Record<string, unknown> = {}
    if (params.search) body.search = params.search
    return rpc<{ locations?: TravelLocation[] }>(SVC, 'ListLocations', body).then(r => r.locations ?? [])
  },

  create: (data: TravelLocationFormData): Promise<TravelLocation> =>
    rpc<{ location: TravelLocation }>(SVC, 'CreateLocation', data).then(r => r.location),

  update: (id: string, data: TravelLocationFormData): Promise<TravelLocation> =>
    rpc<{ location: TravelLocation }>(SVC, 'UpdateLocation', { id, ...data }).then(r => r.location),

  delete: (id: string): Promise<void> =>
    rpc<Record<string, never>>(SVC, 'DeleteLocation', { id }).then(() => undefined),

  createSpot: (locationId: string, name: string, description?: string): Promise<TouristSpot> =>
    rpc<{ spot: TouristSpot }>(SVC, 'CreateSpot', { locationId, name, description }).then(r => r.spot),

  updateSpot: (id: string, name: string, description?: string): Promise<TouristSpot> =>
    rpc<{ spot: TouristSpot }>(SVC, 'UpdateSpot', { id, name, description }).then(r => r.spot),

  deleteSpot: (id: string): Promise<void> =>
    rpc<Record<string, never>>(SVC, 'DeleteSpot', { id }).then(() => undefined),
}
